### PR TITLE
Create valid route names with long namespace names

### PR DESCRIPTION
**Pull request by csrwng**
**What this PR does / why we need it**:
When using a long name for a cluster namespace, the routes that the
control plane operator creates is invalid because a segment of the
domain name has a length > 63 chars.

This commit sets the host name for routes to ensure that the resulting
host name is valid. Modification of the route hostname only happens when
the resulting hostname will be invalid. The name is shortened by
trimming the original name and appending a hash of the entire name,
resulting in repeatable and distinct names every time the name is
shortened.

**Which issue(s) this PR fixes** 
Fixes # [HOSTEDCP-371](https://issues.redhat.com/browse/HOSTEDCP-371)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go [hunk 1]
```diff
@@ -4145,6 +4161,21 @@ func (r *HostedClusterReconciler) defaultAPIPortIfNeeded(ctx context.Context, hc
 	return nil
 }
 
+func (r *HostedClusterReconciler) defaultIngressDomain(ctx context.Context) (string, error) {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4164]: why is this needed? Isn't this available at the HCP API, e.g `HCP.DNSSpec`?

Comment (csrwng) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4164]: This is the ingress domain of the management cluster, not of the HC. We can obtain it from the ingress.config.openshift.io/cluster resource.

**Comment by csrwng** [comment]
/hold
missing shortening of ignition server route

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 1 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4166]: Yeh I think it makes sense for the HC to fetch it as in the PR. But I think it should be passed as API input to the HCP as everything else, why is it using an env variable instead?

Comment (csrwng) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4166]: The reason is that this is an operator-wide setting and nothing specific to a particular HostedCluster. It seemed wrong to pass it as a HostedControlPlane field. If/when it changes, we need to modify all existing HCPs instead of just redeploying the control plane operator.

File: control-plane-operator/controllers/hostedcontrolplane/oauth/route.go [hunk 2]
```diff
@@ -6,16 +6,19 @@ import (
...
+func ReconcileRoute(route *routev1.Route, ownerRef config.OwnerRef, strategy *hyperv1.ServicePublishingStrategy, defaultIngressDomain string) error {
 	ownerRef.ApplyTo(route)
 	if strategy.Route != nil && strategy.Route.Hostname != "" {
 		if route.Annotations == nil {
 			route.Annotations = map[string]string{}
 		}
 		route.Annotations[hyperv1.ExternalDNSHostnameAnnotation] = strategy.Route.Hostname
 		route.Spec.Host = strategy.Route.Hostname
+	} else {
+		route.Spec.Host = util.ShortenRouteHostnameIfNeeded(route.Name, route.Namespace, defaultIngressDomain)
```
Comment (sjenning) [inline control-plane-operator/controllers/hostedcontrolplane/oauth/route.go:21]: not a problem with this PR but... where is the `Private` handling here? :thinking:

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 1 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4166]: ok I see, why is it not a flag then as everything else? to clarify I have no objections to proceed with the env, just trying to understand why we are making a difference here.

Comment (csrwng) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4166]: Because flags introduce compatibility problems that are avoided by env vars. We don't have a good way from the HO to determine whether a flag applies to a particular control plane binary. If we agree, I would prefer to use env vars for future params passed from HO to CPO. If the particular CPO doesn't support it, then it can just ignore it.

File: control-plane-operator/controllers/hostedcontrolplane/oauth/route.go (hunk 2 above)
Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/oauth/route.go:21]: I guess we haven't so far tried to use oauth in a private setup? Seems like we would need to handle it

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go (hunk 1 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go:4166]: @csrwng Make sense but then for long term sustainability it'd be good to try to consolidate and migrate all flags to be envs before cutting a release.
